design-documents/graph-ql/framework/async-mutations.md

```diff
@@ -0,0 +1,281 @@
+        __typename__
+        ... on OperationFailed {
+            failures {
+                message
+                __typename__
+            }
+        }
+    }
+}
+```
+#### Response
+```json
+{
+   "queue":[
+      {
+         "identifier":"25d7d02fa2b73238b098b0254d9ec98a5cd23696",
+         "__typename__":"OperationFailed",
+         "failures":[
+            {
+               "__typename__":"LogicalErrorReport",
+               "message":"No such entity with email example@example.com"
+            }
+         ]
+      }
+   ]
+}
+```
+
+### Retrieving results
+
+Not all of commands require execution results.
+For some operations this is enough to return just an execution status.
+In case the application requires to work with the mutation execution results exists two ways for this. Generic and custom.
+
+#### Generic
+
+Generic way provides a possibility to return with success list of types that were changes during mutation execution.
+Basically, this is a list of updated entities which should be reloaded for application work. The application can use this result as tags for invalidation local cached entities etc.
+Filling of this list is a developer responsibility and does not guarantee by the framework.
+To return information about such entities operation state must implement 
+
+```graphql schema
+type ChangedEntity {
+    type: String!
+    identifiers: [String!]
+}
+
+type PersistOperationInterface {
+    changedEntities: [ChangedEntity!]
+}
+
+type PersistOperationCompleted implements PersistOperationInterface, OperationStateInterface {
+     identifier: String!
+     changedEntities: [ChangedEntity!]
+}
+```
+
+#### Request
+```graphql 
+query {
+    queue($operataionIdentifier: [String!]) {
+        identifier
+        __typename__
+        ... on PersistOperationCompleted {
+            changedEntities {
+                type
+                identifier
+            }
+        }
+    }
+}
+```
+#### Response
+```json
+{
+   "queue":[
+      {
+         "identifier":"3d8d487165a5653733e3a3e8bd46783e46b9a58d",
+         "__typename__":"PersistOperationCompleted",
+         "changedEntities":[
+            {
+               "Cart":[
+                  "29f77a662a7f4d019d5160a9dbe4b83067a75f68"
+               ]
+            }
+         ]
+      }
+   ]
+}
+```
+#### Custom
+
+A developer can specify custom operation status, which may aggregate success result with output information.
+For instance, `CartSaveOperationCompleted` state may, return a corresponding cart as with query.
+```graphql schema
+type CartSaveOperationCompleted implements OperationStateInterface {
+     identifier: String!
+     cart: Cart!
+}
+```
+For such cases, a client should be aware of the operation and result that it expects to receive.
+This is kind of trade-off between generalized approach and numbers of request to the server.
+Such an approach may help to achieve better performance but definitely will cost more in development and testing.
```

Comment:
I would suggest to add recommendation to avoid this option unless needed for specific performance optimization.